rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,528 @@
++  // to be used as an MFA method and if the user is pre-authenticated (not first time login).
++  bool CreatePrivilegedToken = 18 [(gogoproto.jsontag) = "create_token,omitempty"];
+}
+```
+
+**AuthConnectorChallenge**
+```diff
+message MFAAuthenticateChallenge {
+  ...
++  // ProviderChallenge is an an auth provider MFA challenge. If set, the client
++  // will attempt to create an auth request with this connector to acquire a
++  // privileged token as a substitute for local MFA.
++  ProviderChallenge ProviderChallenge = 5;
+}
+
++// ProviderChallenge contains auth connector details for completing a provider
++// MFA challenge.
++message ProviderChallenge {
++  // Type is the auth connector type.
++  string Type = 1;
++  // ID is the auth connector ID.
++  string ID = 2;
++}
+
+message MFAAuthenticateResponse {
+  oneof Response {
+    TOTPResponse TOTP = 2;
+    webauthn.CredentialAssertionResponse Webauthn = 3;
++   PrivilegeTokenResponse Token = 4;
+  }
+}
+```
+
+### Backward Compatibility
+
+It's possible that an old client or proxy version could result in an attempted
+SSO login using an auth connector setup for MFA only. The Auth server will
+check auth requests against the configured auth connector to ensure that these
+login attempts are prevented. Checks will be added to the following endpoints:
+
+- `rpc CreateSAMLAuthRequest`
+- `rpc CreateOIDCAuthRequest`
+- `rpc CreateGithubAuthRequest`
+- `http /v1/saml/requests/validate`
+- `http /v1/oidc/requests/validate`
+- `http /v1/github/requests/validate`
+
+### Audit Events
+
+SSO MFA requests will be tracked through the existing audit events that contain
+`MFADeviceMetadata`. Since SSO as MFA doesn't correspond to actual user MFA
+devices in the backend, this metadata will be picked from the auth connector:
+
+- `DeviceName` - Name of the auth connector
+- `DeviceID` - UUID of the auth connector
+- `DeviceType` - SSO type (`OIDC`, `SAML`, or `Github`)
+
+```proto
+message MFADeviceMetadata {
+  // Name is the user-specified name of the MFA device.
+  string DeviceName = 1 [(gogoproto.jsontag) = "mfa_device_name"];
+  // ID is the UUID of the MFA device generated by Teleport.
+  string DeviceID = 2 [(gogoproto.jsontag) = "mfa_device_uuid"];
+  // Type is the type of this MFA device.
+  string DeviceType = 3 [(gogoproto.jsontag) = "mfa_device_type"];
+}
+```
+
+#### Privilege Tokens
+
+When a privilege token is created, the context of its creation is lost. This
+makes it difficult to tie audit event between the SSO login that creates the
+token and the Per-session MFA certificate issuance that consumes the token.
+
+To amend this, we will store the details of the MFA device used to create the
+privilege token in the backend token resource. This data can then be included
+in the `privilege_token.create` and `mfa_auth_challenge.validate` events.
+
+### Additional considerations
+
+#### Temporary SSO MFA device in backend
```

Comment:
This is interesting.

I think what I like the most about this is that it gives visibility to what kinds of MFA are available for the user. It could maybe help with debugging problems too.

Like you said, nice to have but probably not worth the effort it would take.